File: scripts/ci/assert-rust-llvm-coverage-runner-contract.py
#!/usr/bin/env python3
"""Validate the Buck2-native LLVM Rust coverage runner contract.

This checker is local/static contract evidence only. It validates the target
shape for future Buck2-native LLVM source-based coverage runners. It does not
run tests, generate coverage reports, invoke llvm-profdata/llvm-cov, post
statuses, mutate branch protection, or prove P0.0 / Phase-0 exit authority.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from typing import Any

DEFAULT_SPEC = Path("specs/rust-llvm-coverage-runner-contract.json")
FALSE_CLAIMS = (
    "coverage_report_generated",
    "coverage_budget_enforced",
    "status_mutation_performed",
    "protected_branch_authority_proven",
    "live_required_context_execution_proven",
    "p0_0_green",
    "phase0_complete",
    "production_ready",
    "hyperscaler_grade",
)
REQUIRED_OFFICIAL_URLS = (
    "https://doc.rust-lang.org/rustc/instrument-coverage.html",
    "https://clang.llvm.org/docs/SourceBasedCodeCoverage.html",
    "https://buck2.build/docs/users/commands/",
)
REQUIRED_EVIDENCE_FIELDS = (
    "Buck2 target",
    "Buck2 Build ID",
    "report path",
    "changed-file delta",
    "excluded generated paths",
    "coverage budget result",
)
REQUIRED_TOOLCHAIN_TOOLS = ("rustc", "llvm-profdata", "llvm-cov")
RAW_CARGO_RE = re.compile(r"(^|[;&|(`]|\s)cargo\s+([a-z0-9_-]+)(\s|$)", re.IGNORECASE)


def load_json(path: Path) -> dict[str, Any]:
    with path.open() as fh:
        data = json.load(fh)
    if not isinstance(data, dict):
        raise TypeError(f"{path}: expected JSON object")
    return data


def strings(value: Any) -> list[str]:
    if isinstance(value, list):
        return [item for item in value if isinstance(item, str)]
    return []


def official_urls(spec: dict[str, Any]) -> set[str]:
    refs = spec.get("_meta", {}).get("official_references", [])
    urls: set[str] = set()
    if isinstance(refs, list):
        for ref in refs:
            if isinstance(ref, dict) and isinstance(ref.get("url"), str):
                urls.add(ref["url"])
    return urls


def validate(spec: dict[str, Any]) -> dict[str, Any]:
    failures: list[str] = []
    boundary = spec.get("claim_boundary") if isinstance(spec.get("claim_boundary"), dict) else {}
    pipeline = spec.get("coverage_pipeline") if isinstance(spec.get("coverage_pipeline"), dict) else {}
    instrumentation = pipeline.get("instrumentation") if isinstance(pipeline.get("instrumentation"), dict) else {}
    merge = pipeline.get("profile_merge") if isinstance(pipeline.get("profile_merge"), dict) else {}
    export = pipeline.get("report_export") if isinstance(pipeline.get("report_export"), dict) else {}
    toolchain = (
        pipeline.get("toolchain_requirements")
        if isinstance(pipeline.get("toolchain_requirements"), dict)
        else {}
    )
    smoke = pipeline.get("smoke_coverage") if isinstance(pipeline.get("smoke_coverage"), dict) else {}

    if boundary.get("coverage_runner_contract_proven") is not True:
        failures.append("coverage_runner_contract_not_proven")
    for claim in FALSE_CLAIMS:
        if boundary.get(claim) is not False:
            failures.append(f"forbidden_true_or_missing_claim_{claim}")

    if pipeline.get("canonical_surface") != "Buck2-native LLVM source-based coverage":
        failures.append("missing_buck2_native_llvm_canonical_surface")
    noncanonical = pipeline.get("noncanonical_surface")
    if not isinstance(noncanonical, str) or "Tarpaulin" not in noncanonical or "not required CI/PR" not in noncanonical:
        failures.append("tarpaulin_noncanonical_boundary_missing")
    if pipeline.get("buck2_contract_target") != "//:rust-llvm-coverage-runner-contract-check":
        failures.append("wrong_buck2_contract_target")
    if pipeline.get("future_runner_authority") != "trusted cloud-ci/oya-ci Buck2 target inventory":
        failures.append("wrong_future_runner_authority")

    if toolchain.get("ambient_path_llvm_tools_required") is not False:
        failures.append("ambient_path_llvm_tools_not_forbidden")
    if "rustup llvm-tools component" not in str(toolchain.get("local_smoke_llvm_tools_source", "")):
        failures.append("missing_rustup_llvm_tools_source")
    if "trusted cloud-ci/oya-ci Buck2 toolchain inventory" not in str(
        toolchain.get("live_runner_llvm_tools_source", "")
    ):
        failures.append("missing_live_runner_toolchain_inventory_source")
    if "rustc --print sysroot" not in str(toolchain.get("host_tool_path_derivation", "")):
        failures.append("missing_sysroot_tool_path_derivation")
    if "pin" not in str(toolchain.get("pinning_requirement", "")).lower():
        failures.append("missing_toolchain_pinning_requirement")
    required_tools = set(strings(toolchain.get("required_tools")))
    for tool in REQUIRED_TOOLCHAIN_TOOLS:
        if tool not in required_tools:
            failures.append(f"missing_required_toolchain_tool_{tool.replace('-', '_')}")

    if smoke.get("buck2_smoke_target") != "//:rust-llvm-coverage-smoke-check":
        failures.append("missing_buck2_coverage_smoke_target")
    if smoke.get("smoke_script") != "scripts/ci/run-rust-llvm-coverage-smoke.py":
        failures.append("missing_coverage_smoke_script")
    if smoke.get("fixture") != "specs/fixtures/rust-llvm-coverage-smoke/branchy.rs":
        failures.append("missing_coverage_smoke_fixture")
    if smoke.get("fixture_report_generated") is not True:
        failures.append("fixture_report_generation_not_recorded")
    if smoke.get("production_coverage_report_generated") is not False:
        failures.append("production_coverage_false_boundary_missing")
    if "none" not in str(smoke.get("budget_authority", "")).lower():
        failures.append("smoke_budget_authority_not_none")

    evidence_fields = strings(pipeline.get("required_evidence_fields"))
    for field in REQUIRED_EVIDENCE_FIELDS:
        if field not in evidence_fields:
            failures.append(f"missing_required_evidence_field_{field.replace(' ', '_').replace('-', '_').lower()}")

    if instrumentation.get("rustc_flag") != "rustc -C instrument-coverage":
        failures.append("missing_instrument_coverage_flag")
    if instrumentation.get("buck2_rust_rule_field") != "rustc_flags":
        failures.append("missing_buck2_rustc_flags_field")
    if instrumentation.get("profile_env_var") != "LLVM_PROFILE_FILE":
        failures.append("missing_llvm_profile_file_env")
    template = instrumentation.get("profile_template")
    if not isinstance(template, str) or "%m-%p" not in template or not template.endswith(".profraw"):
        failures.append("missing_profile_collision_guard_or_profraw_template")
    if instrumentation.get("profile_collision_guard") != "%m-%p":
        failures.append("missing_profile_collision_guard")
    if instrumentation.get("profile_extension") != ".profraw":
        failures.append("missing_profraw_extension")

    if merge.get("tool") != "llvm-profdata":
        failures.append("missing_llvm_profdata_tool")
    if merge.get("operation") != "merge":
        failures.append("missing_profdata_merge_operation")
    if merge.get("mode") != "-sparse":
        failures.append("missing_sparse_profdata_merge_mode")
    if "*.profraw" not in str(merge.get("input_glob")):
        failures.append("missing_profraw_merge_input_glob")
    if not str(merge.get("output", "")).endswith(".profdata"):
        failures.append("missing_profdata_output")

    if export.get("tool") != "llvm-cov":
        failures.append("missing_llvm_cov_tool")
    formats = set(strings(export.get("formats")))
    for fmt in ("text", "html", "json"):
        if fmt not in formats:
            failures.append(f"missing_llvm_cov_{fmt}_format")
    budget = export.get("delta_budget") if isinstance(export.get("delta_budget"), dict) else {}
    if budget.get("changed_files_line_coverage_minimum") != 80:
        failures.append("wrong_changed_file_delta_budget")
    if budget.get("kernel_domain_absolute_line_coverage_minimum") != 70:
        failures.append("wrong_kernel_domain_budget")
    if budget.get("generated_code_excluded") is not True:
        failures.append("generated_code_exclusion_missing")

    forbidden = "\n".join(strings(pipeline.get("forbidden_authority")))
    if "Tarpaulin as canonical monorepo coverage evidence" not in forbidden:
        failures.append("missing_tarpaulin_forbidden_authority")
    if "candidate-authored target inventory as coverage authority" not in forbidden:
        failures.append("missing_candidate_inventory_forbidden_authority")

    urls = official_urls(spec)
    for url in REQUIRED_OFFICIAL_URLS:
        if url not in urls:
            failures.append(f"missing_official_reference_{url.rsplit('/', 2)[-2] if url.endswith('/') else url.rsplit('/', 1)[-1]}")

    rendered = json.dumps(spec, sort_keys=True)
    if RAW_CARGO_RE.search(rendered):
        failures.append("raw_cargo_command_present_in_contract")

    automated_chain = "\n".join(strings(spec.get("automated_chain")))
    if "buck2 build //:rust-llvm-coverage-smoke-check" not in automated_chain:
        failures.append("missing_smoke_target_in_automated_chain")

    return {
        "authority_boundary": "local/static coverage runner contract only; no coverage report generated and no live required-context authority proven",
        "coverage_runner_contract_proven": not failures,
        "coverage_report_generated": False,
        "coverage_budget_enforced": False,
        "status_mutation_performed": False,
        "protected_branch_authority_proven": False,
        "live_required_context_execution_proven": False,
        "p0_0_green": False,
        "phase0_complete": False,
        "production_ready": False,
        "hyperscaler_grade": False,
        "verdict": "PASS" if not failures else "FAIL",
        "failures": sorted(set(failures)),
    }


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--spec", default=str(DEFAULT_SPEC), help="coverage runner contract spec path")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    result = validate(load_json(Path(args.spec)))
    rendered = json.dumps(result, sort_keys=True)
    if args.json or result["verdict"] == "PASS":
        print(rendered)
    else:
        print(rendered, file=sys.stderr)
    return 0 if result["verdict"] == "PASS" else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/ci/assert-rust-llvm-coverage-runner-contract.rs
//! Rust LLVM coverage-runner contract checker.
//!
//! Local/static contract evidence only: validates the Buck2-native LLVM
//! source-based coverage target shape, explicit Tarpaulin non-authority, and
//! non-claim boundary. It does not run coverage, post statuses, mutate branch
//! protection, or prove live required-context authority.

use std::env;
use std::fs;

const DEFAULT_SPEC: &str = "specs/rust-llvm-coverage-runner-contract.json";
const FALSE_CLAIMS: &[&str] = &[
    "coverage_report_generated",
    "coverage_budget_enforced",
    "status_mutation_performed",
    "protected_branch_authority_proven",
    "live_required_context_execution_proven",
    "p0_0_green",
    "phase0_complete",
    "production_ready",
    "hyperscaler_grade",
];
const REQUIRED_OFFICIAL_URLS: &[&str] = &[
    "https://doc.rust-lang.org/rustc/instrument-coverage.html",
    "https://clang.llvm.org/docs/SourceBasedCodeCoverage.html",
    "https://buck2.build/docs/users/commands/",
];
const REQUIRED_EVIDENCE_FIELDS: &[&str] = &[
    "Buck2 target",
    "Buck2 Build ID",
    "report path",
    "changed-file delta",
    "excluded generated paths",
    "coverage budget result",
];
const REQUIRED_TOOLCHAIN_TOOLS: &[&str] = &["rustc", "llvm-profdata", "llvm-cov"];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Evaluation {
    pub verdict: String,
    pub failures: Vec<String>,
}

fn json_escape(input: &str) -> String {
    input
        .chars()
        .flat_map(|ch| match ch {
            '\\' => "\\\\".chars().collect::<Vec<_>>(),
            '"' => "\\\"".chars().collect::<Vec<_>>(),
            '\n' => "\\n".chars().collect::<Vec<_>>(),
            '\r' => "\\r".chars().collect::<Vec<_>>(),
            '\t' => "\\t".chars().collect::<Vec<_>>(),
            _ => vec![ch],
        })
        .collect()
}

fn json_string(value: &str) -> String {
    format!("\"{}\"", json_escape(value))
}

fn compact_json_text(input: &str) -> String {
    input.chars().filter(|ch| !ch.is_whitespace()).collect()
}

fn contains_json_string(text: &str, value: &str) -> bool {
    text.contains(&json_string(value))
}

fn has_bool(text: &str, key: &str, value: bool) -> bool {
    compact_json_text(text).contains(&format!(
        "\"{}\":{}",
        key,
        if value { "true" } else { "false" }
    ))
}

fn has_string_value(text: &str, key: &str, value: &str) -> bool {
    let key_token = json_string(key);
    let value_token = json_string(value);
    let mut cursor = 0usize;
    while let Some(relative) = text[cursor..].find(&key_token) {
        cursor += relative + key_token.len();
        let mut rest = &text[cursor..];
        rest = rest.trim_start();
        let Some(after_colon) = rest.strip_prefix(':') else {
            continue;
        };
        if after_colon.trim_start().starts_with(&value_token) {
            return true;
        }
    }
    false
}

fn require(condition: bool, failures: &mut Vec<String>, failure: &str) {
    if !condition {
        failures.push(failure.to_owned());
    }
}

fn raw_cargo_command_present(text: &str) -> bool {
    for raw_line in text.lines() {
        let line = raw_line.to_ascii_lowercase();
        let mut index = 0usize;
        while let Some(pos) = line[index..].find("cargo") {
            let start = index + pos;
            let before = line[..start].chars().next_back();
            let after = start + "cargo".len();
            let before_ok = before
                .map(|ch| ch.is_whitespace() || matches!(ch, ';' | '&' | '|' | '(' | '`'))
                .unwrap_or(true);
            let after_ok = line[after..]
                .chars()
                .next()
                .map(|ch| ch.is_whitespace())
                .unwrap_or(false);
            if before_ok && after_ok {
                let mut cursor = after;
                while line[cursor..]
                    .chars()
                    .next()
                    .map(|ch| ch.is_whitespace())
                    .unwrap_or(false)
                {
                    cursor += line[cursor..].chars().next().unwrap().len_utf8();
                }
                let command_start = cursor;
                while cursor < line.len() {
                    let ch = line[cursor..].chars().next().unwrap();
                    if !(ch.is_ascii_alphanumeric() || ch == '_' || ch == '-') {
                        break;
                    }
                    cursor += ch.len_utf8();
                }
                let terminated = line[cursor..]
                    .chars()
                    .next()
                    .map(|ch| ch.is_whitespace())
                    .unwrap_or(true);
                if terminated && cursor > command_start {
                    return true;
                }
            }
            index = after;
        }
    }
    false
}

pub fn evaluate_text(spec: &str) -> Evaluation {
    let mut failures = Vec::new();

    require(
        has_bool(spec, "coverage_runner_contract_proven", true),
        &mut failures,
        "coverage_runner_contract_not_proven",
    );
    for claim in FALSE_CLAIMS {
        require(
            has_bool(spec, claim, false),
            &mut failures,
            &format!("forbidden_true_or_missing_claim_{claim}"),
        );
    }

    require(
        has_string_value(
            spec,
            "canonical_surface",
            "Buck2-native LLVM source-based coverage",
        ),
        &mut failures,
        "missing_buck2_native_llvm_canonical_surface",
    );
    require(
        spec.contains("Tarpaulin") && spec.contains("not required CI/PR"),
        &mut failures,
        "tarpaulin_noncanonical_boundary_missing",
    );
    require(
        has_string_value(
            spec,
            "buck2_contract_target",
            "//:rust-llvm-coverage-runner-contract-check",
        ),
        &mut failures,
        "wrong_buck2_contract_target",
    );
    require(
        has_string_value(
            spec,
            "future_runner_authority",
            "trusted cloud-ci/oya-ci Buck2 target inventory",
        ),
        &mut failures,
        "wrong_future_runner_authority",
    );

    require(
        has_bool(spec, "ambient_path_llvm_tools_required", false),
        &mut failures,
        "ambient_path_llvm_tools_not_forbidden",
    );
    require(
        spec.contains("rustup llvm-tools component"),
        &mut failures,
        "missing_rustup_llvm_tools_source",
    );
    require(
        spec.contains("trusted cloud-ci/oya-ci Buck2 toolchain inventory"),
        &mut failures,
        "missing_live_runner_toolchain_inventory_source",
    );
    require(
        spec.contains("rustc --print sysroot"),
        &mut failures,
        "missing_sysroot_tool_path_derivation",
    );
    require(
        spec.to_ascii_lowercase().contains("pin"),
        &mut failures,
        "missing_toolchain_pinning_requirement",
    );
    for tool in REQUIRED_TOOLCHAIN_TOOLS {
        require(
            contains_json_string(spec, tool),
            &mut failures,
            &format!("missing_required_toolchain_tool_{}", tool.replace('-', "_")),
        );
    }

    require(
        has_string_value(
            spec,
            "buck2_smoke_target",
            "//:rust-llvm-coverage-smoke-check",
        ),
        &mut failures,
        "missing_buck2_coverage_smoke_target",
    );
    require(
        has_string_value(
            spec,
            "smoke_script",
            "scripts/ci/run-rust-llvm-coverage-smoke.rs",
        ),
        &mut failures,
        "missing_coverage_smoke_script",
    );
    require(
        has_string_value(
            spec,
            "fixture",
            "specs/fixtures/rust-llvm-coverage-smoke/branchy.rs",
        ),
        &mut failures,
        "missing_coverage_smoke_fixture",
    );
    require(
        has_bool(spec, "fixture_report_generated", true),
        &mut failures,
        "fixture_report_generation_not_recorded",
    );
    require(
        has_bool(spec, "production_coverage_report_generated", false),
        &mut failures,
        "production_coverage_false_boundary_missing",
    );
    require(
        spec.to_ascii_lowercase().contains("none"),
        &mut failures,
        "smoke_budget_authority_not_none",
    );

    for field in REQUIRED_EVIDENCE_FIELDS {
        require(
            contains_json_string(spec, field),
            &mut failures,
            &format!(
                "missing_required_evidence_field_{}",
                field
                    .replace(' ', "_")
                    .replace('-', "_")
                    .to_ascii_lowercase()
            ),
        );
    }

    require(
        has_string_value(spec, "rustc_flag", "rustc -C instrument-coverage"),
        &mut failures,
        "missing_instrument_coverage_flag",
    );
    require(
        has_string_value(spec, "buck2_rust_rule_field", "rustc_flags"),
        &mut failures,
        "missing_buck2_rustc_flags_field",
    );
    require(
        has_string_value(spec, "profile_env_var", "LLVM_PROFILE_FILE"),
        &mut failures,
        "missing_llvm_profile_file_env",
    );
    require(
        spec.contains("%m-%p") && spec.contains(".profraw"),
        &mut failures,
        "missing_profile_collision_guard_or_profraw_template",
    );
    require(
        has_string_value(spec, "profile_collision_guard", "%m-%p"),
        &mut failures,
        "missing_profile_collision_guard",
    );
    require(
        has_string_value(spec, "profile_extension", ".profraw"),
        &mut failures,
        "missing_profraw_extension",
    );

    require(
        has_string_value(spec, "tool", "llvm-profdata"),
        &mut failures,
        "missing_llvm_profdata_tool",
    );
    require(
        has_string_value(spec, "operation", "merge"),
        &mut failures,
        "missing_profdata_merge_operation",
    );
    require(
        has_string_value(spec, "mode", "-sparse"),
        &mut failures,
        "missing_sparse_profdata_merge_mode",
    );
    require(
        spec.contains("*.profraw"),
        &mut failures,
        "missing_profraw_merge_input_glob",
    );
    require(
        spec.contains(".profdata"),
        &mut failures,
        "missing_profdata_output",
    );

    require(
        has_string_value(spec, "tool", "llvm-cov"),
        &mut failures,
        "missing_llvm_cov_tool",
    );
    for fmt in ["text", "html", "json"] {
        require(
            contains_json_string(spec, fmt),
            &mut failures,
            &format!("missing_llvm_cov_{fmt}_format"),
        );
    }
    require(
        compact_json_text(spec).contains("\"changed_files_line_coverage_minimum\":80"),
        &mut failures,
        "wrong_changed_file_delta_budget",
    );
    require(
        compact_json_text(spec).contains("\"kernel_domain_absolute_line_coverage_minimum\":70"),
        &mut failures,
        "wrong_kernel_domain_budget",
    );
    require(
        has_bool(spec, "generated_code_excluded", true),
        &mut failures,
        "generated_code_exclusion_missing",
    );

    require(
        spec.contains("Tarpaulin as canonical monorepo coverage evidence"),
        &mut failures,
        "missing_tarpaulin_forbidden_authority",
    );
    require(
        spec.contains("candidate-authored target inventory as coverage authority"),
        &mut failures,
        "missing_candidate_inventory_forbidden_authority",
    );
    for url in REQUIRED_OFFICIAL_URLS {
        let suffix = if url.ends_with('/') {
            url.trim_end_matches('/').rsplit('/').next().unwrap_or(url)
        } else {
            url.rsplit('/').next().unwrap_or(url)
        };
        require(
            contains_json_string(spec, url),
            &mut failures,
            &format!("missing_official_reference_{suffix}"),
        );
    }
    require(
        !raw_cargo_command_present(spec),
        &mut failures,
        "raw_cargo_command_present_in_contract",
    );
    require(
        spec.contains("buck2 build //:rust-llvm-coverage-smoke-check"),
        &mut failures,
        "missing_smoke_target_in_automated_chain",
    );

    failures.sort();
    failures.dedup();
    Evaluation {
        verdict: if failures.is_empty() { "PASS" } else { "FAIL" }.to_owned(),
        failures,
    }
}

pub fn evaluate_file(path: &str) -> Evaluation {
    match fs::read_to_string(path) {
        Ok(text) => evaluate_text(&text),
        Err(error) => Evaluation {
            verdict: "FAIL".to_owned(),
            failures: vec![format!("spec_read_failed:{error}")],
        },
    }
}

fn render_json(evaluation: &Evaluation) -> String {
    let failures = evaluation
        .failures
        .iter()
        .map(|failure| json_string(failure))
        .collect::<Vec<_>>()
        .join(",");
    format!(
        "{{\"authority_boundary\":\"local/static coverage runner contract only; no coverage report generated and no live required-context authority proven\",\"coverage_budget_enforced\":false,\"coverage_report_generated\":false,\"coverage_runner_contract_proven\":{},\"failures\":[{}],\"hyperscaler_grade\":false,\"live_required_context_execution_proven\":false,\"p0_0_green\":false,\"phase0_complete\":false,\"production_ready\":false,\"protected_branch_authority_proven\":false,\"status_mutation_performed\":false,\"verdict\":{}}}",
        if evaluation.failures.is_empty() {
            "true"
        } else {
            "false"
        },
        failures,
        json_string(&evaluation.verdict),
    )
}

fn parse_args() -> (String, bool) {
    let mut spec = DEFAULT_SPEC.to_owned();
    let mut json = false;
    let mut args = env::args().skip(1);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--spec" => spec = args.next().unwrap_or_else(|| spec.clone()),
            "--json" => json = true,
            _ => {}
        }
    }
    (spec, json)
}

fn main() {
    let (spec, json) = parse_args();
    let evaluation = evaluate_file(&spec);
    let rendered = render_json(&evaluation);
    if json || evaluation.verdict == "PASS" {
        println!("{rendered}");
    } else {
        eprintln!("{rendered}");
    }
    if evaluation.verdict != "PASS" {
        std::process::exit(1);
    }
}
